src/main/java/org/jvnet/hudson/test/LoggerRule.java

```diff
@@ -0,0 +1,129 @@
+import java.util.Map;
+import java.util.logging.ConsoleHandler;
+import java.util.logging.Formatter;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
+import java.util.logging.SimpleFormatter;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.rules.ExternalResource;
+import org.junit.rules.RuleChain;
+
+/**
+ * A test rule which allows you to easily enable one or more loggers for the duration of a test.
+ * Call {@link #record(Class, Level)} or another overload for the rule to take effect.
+ * <p>To capture messages during Jenkins startup,
+ * you may compose this with a {@link JenkinsRule} using a {@link RuleChain};
+ * or use as a {@link ClassRule}.
+ */
+public class LoggerRule extends ExternalResource {
+
+    private final RingBufferLogHandler ringHandler = new RingBufferLogHandler();
+    private final ConsoleHandler consoleHandler = new ConsoleHandler();
+    private final Map<Logger,Level> loggers = new HashMap<Logger,Level>();
+
+    /**
+     * Initializes the rule, by default not recording anything.
+     */
+    public LoggerRule() {
+        ringHandler.setLevel(Level.ALL);
+        consoleHandler.setLevel(Level.ALL);
+    }
+
+    /**
+     * Start listening to a logger.
+     * Might be called in a {@link Rule} initializer, to apply to all test cases in a suite;
+     * or only at the start of selected test cases.
+     * @param logger some logger
+     * @param level something between {@link Level#CONFIG} and {@link Level#ALL};
+     *              using {@link Level#INFO} or above is typically senseless,
+     *              since Java will by default log everything at such levels anyway;
+     *              unless you wish to inspect visible {@link #getRecords},
+     *              or wish to <em>suppress</em> console log output for some logger
+     * @return this rule, for convenience
+     */
+    public LoggerRule record(Logger logger, Level level) {
+        loggers.put(logger, logger.getLevel());
+        logger.setLevel(level);
+        logger.addHandler(ringHandler);
+        logger.addHandler(consoleHandler);
+        return this;
+    }
+
+    /**
+     * Same as {@link #record(Logger, Level) but calls {@link Logger#getLogger(String)} for you first.
+     */
+    public LoggerRule record(String name, Level level) {
+        return record(Logger.getLogger(name), level);
+    }
+    
+    /**
+     * Same as {@link #record(String, Level) but calls {@link Class#getName()} for you first.
+     */
+    public LoggerRule record(Class<?> clazz, Level level) {
+        return record(clazz.getName(), level);
+    }
+
+    /**
+     * Obtains all log records collected so far during this test case.
+     */
+    public List<LogRecord> getRecords() {
+        return ringHandler.getView();
+    }
+
+    /**
+     * Like {@link #getRecords} but applies {@link SimpleFormatter#format(LogRecord)}.
+     */
+    public List<String> getRecordsFormatted() {
```

Comment:
so with the normal handlers in java.util.logging the message is computed on the call stack in the logger.log(...) method.  Thus if you pass in non idempotent Objects for the formatter (`Log.log(Level.INFO, "{0}, {1}" new Object[mutable1, mutable2]`) then the message is created at a time that represents the state of those mutable as seen in the caller - if you are logging mutable state that is being changed by multiple threads then you are SOOL.

Now - some logging handlers/frameworks allow for asynchronous logging - such that the creation of the message is decoupled from the caller of the log method.  such that the state of the mutable objects will be different at the time the message is created.

e.g.  with a synchronous logging handler...

```java
class Foo {
  public String s;
  public String toString() { return s }
}
// some class
Foo f = new Foo();
f.s = "jesse"
Logger.log(Level.INFO, "Hello from {0}", f);
f.s = "jglick";
```

so that will with synchronous logging (all handlers etc) print  "Hello from jesse".  
However when you then test the method to see if "Hello from jesse" was printed - you will return false (as its now formatted asynchrously from the call to log).